### PR TITLE
Run vivado synth/impl in one instance for ooc designs

### PR DESCRIPTION
**Pull request by reillymck**
There is a new tool wrapper, vivado_full, will write one tcl script to run both synthesis and implementation. This way you don't have the time exiting/starting vivado holding up your flow. If only implementation is out of date, then it only runs implementation (tool outputs for synthesis and implementation are still both produced and the normal file path). Otherwise, it runs both. Also has some fixes so that out-of-context runs correctly for vivado. Also, out-of-context only needs to be specified for synthesis for this flow.

### CODE REVIEW

File: scripts/run_design.py [hunk 1]
```diff
@@ -70,10 +71,11 @@ def main():
         "cmp": ToolType.CMP,
         "reverse": ToolType.REVERSE,
     }
+    print(args)
     print(flow_args[ToolType.SYNTH])
     for arg_name, enum in flow_args_map.items():
-        flow_args[enum] = getattr(args, arg_name)
-
+        flow_args[enum] = getattr(args, arg_name)[0]
```
[the remark this answers is not in the document]
Comment (reillymck) [reply scripts/run_design.py:77]: See above.

File: scripts/run_design.py [hunk 2]
```diff
@@ -40,11 +40,12 @@ def main():
...
-    parser.add_argument("--impl", help="Implementation args", nargs=1, type=str, default="")
-    parser.add_argument("--map", help="Mapping args", nargs=1, type=str, default="")
-    parser.add_argument("--cmp", help="Comparison args", nargs=1, type=str, default="")
-    parser.add_argument("--err", help="Error flow args", nargs=1, type=str, default="")
+    parser.add_argument("--synth", help="Synthesis args", nargs=1, type=str, default=[""])
+    parser.add_argument("--impl", help="Implementation args", nargs=1, type=str, default=[""])
+    parser.add_argument("--map", help="Mapping args", nargs=1, type=str, default=[""])
+    parser.add_argument("--cmp", help="Comparison args", nargs=1, type=str, default=[""])
+    parser.add_argument("--reverse", help="Reverse args", nargs=1, type=str, default=[""])
+    parser.add_argument("--err", help="Error flow args", nargs=1, type=str, default=[""])
```
[the remark this answers is not in the document]
Comment (reillymck) [reply scripts/run_design.py:48]: Basically, when you use "nargs" your return type is a list. But the default was an empty string. So, I changed it to be consistent. However, shlex fails on a list (it expects a string). 
See this: https://stackoverflow.com/questions/25343868/python-argparse-default-argument-stored-as-string-not-list
From the argparse [docs](https://docs.python.org/3/library/argparse.html#nargs):
### nargs
ArgumentParser objects usually associate a single command-line argument with a single action to be taken. The nargs keyword argument associates a different number of command-line arguments with a single action. The supported values are:

N (an integer). N arguments from the command line will be gathered together into a list. For example:
```
parser = argparse.ArgumentParser()
parser.add_argument('--foo', nargs=2)
parser.add_argument('bar', nargs=1)
parser.parse_args('c --foo a b'.split())
Namespace(bar=['c'], foo=['a', 'b'])
```
Note that nargs=1 produces a list of one item. This is different from the default, in which the item is produced by itself.

[the remark this answers is not in the document]
Comment (jgoeders) [reply scripts/run_design.py:48]: Should we just get rid of the nargs=1?  I think I put it there by mistake before I realized that it produced a list.

[the remark this answers is not in the document]
Comment (reillymck) [reply scripts/run_design.py:48]: Yes, I will do that.

**Comment by jgoeders** [comment]
@reillymck Is this ready for review?

**Pull request merged**